### PR TITLE
Fix problem about delete multi table , select into and select lock statement.

### LINKED ISSUE #7978: When the table is not sharding , in 5.0.0, SQL parsing is wrong, but it is correct before
## Bug Report

I have tested some SQL  between older and newest version , and found some SQL parsing errors but work wells before.

### Which version of ShardingSphere did you use?

- **Older version**
    5.0.0-RC1 ( 2020-10-26T16 master)

- **Newest version**
    5.0.0-RC1 ( 2020-10-30T15 master)

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?
 ShardingSphere-Proxy
### Expected behavior
parse correct 
### Actual behavior
parse error
### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.
- **Results in 5.0.0.RC1-older and 5.0.0.RC1-newest**

```
ShardingProxy(5.0.0.RC1-older)No Sharding
deleteMultiTable[hasShardingKey:none];	 Support:true;	SQL: delete customer, customer_email from customer , customer_email where customer.status = 3;
ds_00|delete customer, customer_email from customer , customer_email where customer.status = 3|


ShardingProxy(5.0.0.RC1-newest)No Sharding
deleteMultiTable[hasShardingKey:none];	 Support:false;	SQL: delete customer, customer_email from customer , customer_email where customer.status = 3;
java.sql.SQLException: 2Unknown exception: [null]


ShardingProxy(5.0.0.RC1-older)No Sharding
selectIntoVarAfterFrom[hasShardingKey:none];	 Support:true;	SQL: select * from customer into @myvar;
ds_00|select * from customer into @myvar|

ShardingProxy(5.0.0.RC1-newest)No Sharding
selectIntoVarAfterFrom[hasShardingKey:none];	 Support:false;	SQL: select * from customer into @myvar;
java.sql.SQLException: 2Unknown exception: [org.apache.shardingsphere.sql.parser.sql.common.value.identifier.IdentifierValue cannot be cast to org.apache.shardingsphere.sql.parser.sql.dialect.statement.mysql.dml.MySQLSelectStatement]

```
-  **5.0.0-RC1 config**

```yaml
schemaName: spsqltest_nosharding
#
dataSourceCommon:
  username: root
  password: root135
  connectionTimeoutMilliseconds: 30000
  idleTimeoutMilliseconds: 60000
  maxLifetimeMilliseconds: 1800000
  maxPoolSize: 50
  minPoolSize: 1
  maintenanceIntervalMilliseconds: 30000
#
dataSources:
  ds_00:
    url: jdbc:mysql://127.0.0.1:3306/spsqltest_nosharding?serverTimezone=UTC&useSSL=false

#
rules:
- !SHARDING
  defaultDatabaseStrategy:
    none:
  defaultTableStrategy:
    none:
```

-  **SQL for creating tables**
create before shardingsphere start
```sql
-- @title:createTableCustomer
CREATE TABLE `customer` (
  `id` bigint(20) NOT NULL COMMENT 'primary key',
  `party_id` bigint(20) NOT NULL COMMENT 'user id',
   `status` smallint(6) DEFAULT NULL COMMENT 'status 1 enable 0 disable',
  PRIMARY KEY (`id`),
  KEY `party_id_index` (`party_id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COMMENT='customer';

-- @title:createTableCustomerEmail
CREATE TABLE customer_email (
  `id` bigint(20) NOT NULL COMMENT 'primary key',
  `party_id` bigint(20) NOT NULL COMMENT 'user id',
   `status` smallint(6) DEFAULT NULL COMMENT 'status 1 enable 0 disable',
  PRIMARY KEY (`id`),
  KEY `party_id_index` (`party_id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COMMENT='customer email';
```
- **SQL for testing**

```sql
-- @title:deleteMultiTable,hasShardingKey:none
delete customer, customer_email from customer , customer_email where customer.status = 3;

-- @title:selectIntoVarAfterFrom,hasShardingKey:none
select * from customer into @myvar;

```

### PR DESCRIPTION
**Pull request by lwtdev**
Fixes #7978

* **Fix parse problem in delete multi table statement**
  * Added missing logic for `selectWithInto` when visit select.
  * Supported table assert for `DeleteStatementTestCase`
  * Added test cases for select-into statements; case `select_into_out_file_with_fields_and_escaped` and
    `select_into_with_lock_before_into` are supported in MySQL 8.0 but not in MySQL 5.7.
* **Fix parse problem in  select into statement**
  * Table names should split with `COMMA_ (',')` in `multipleTablesClause`.
  * Added test cases for delete multi table statements.
* **Support `for xxx of table_xxx` for select lock statement in MySQL**
  * Fix definition of `lockClause`
  * Supported lock assert for `SelectStatementTestCase`, add test cases for select lock statement
  * For table rewriting, change definition of `LockSegment`, add `extractTablesFromLock` in `TableExtractor`

### CODE REVIEW

**Comment by lwtdev** [comment]
@jingshanglu  Hi, I have just fix the definition of `lockClause` in this pull request.

**Comment by tristaZero** [comment]
Hi @lwtdev Glad to see your first PR for this community! 
Bug-fix and UT are exactly what we expected. :)

@strongduanmu  @kimmking  @SteNicholas Could any committer give this PR a review? (As I see your names in the relevant ISSUE).

**Comment by strongduanmu** [comment]
> Hi @lwtdev Glad to see your first PR for this community!
> Bug-fix and UT are exactly what we expected. :)
> 
> @strongduanmu @kimmking @SteNicholas Could any committer give this PR a review? (As I see your names in the relevant ISSUE).

@tristaZero I will review this pr. 😀

File: shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/jaxb/cases/domain/segment/impl/lock/ExpectedLockClause.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ * Expected lock clause.
+ */
+@Getter
+@Setter
+public final class ExpectedLockClause extends AbstractExpectedSQLSegment {
+
+    @XmlElement(name = "table")
+    private final List<ExpectedSimpleTable> tables = new LinkedList<>();
+
+}
```
Comment (jingshanglu) [inline shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/jaxb/cases/domain/segment/impl/lock/ExpectedLockClause.java:39]: Delete extra blank lines.

**Review by jingshanglu — CHANGES_REQUESTED** [review]

**Comment by jingshanglu** [comment]
@lwtdev  good job,but there are some format problems, please modify it.

**Comment by jingshanglu** [comment]
@lwtdev Now, there is no UT for expr, are you interested in fix it? The relevant issue is #7880

File: shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/jaxb/cases/domain/segment/impl/lock/ExpectedLockClause.java (hunk 1 above)
Comment (lwtdev) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/main/java/org/apache/shardingsphere/test/sql/parser/parameterized/jaxb/cases/domain/segment/impl/lock/ExpectedLockClause.java:39]: OK, I will remove this blank line.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/test/java/org/apache/shardingsphere/sql/parser/sql/common/extractor/TableExtractorTest.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        LockSegment lockSegment = new LockSegment(108, 154);
+        selectStatement.setLock(lockSegment);
+        List<SimpleTableSegment> tables = new LinkedList<>();
+        lockSegment.setForTables(tables);
+        tables.add(new SimpleTableSegment(122, 128, new IdentifierValue("t_order")));
+        tables.add(new SimpleTableSegment(143, 154, new IdentifierValue("t_order_item")));
+
+        tableExtractor.extractTablesFromSelect(selectStatement);
+
+        assertNotNull(tableExtractor.getRewriteTables());
```
[the remark this answers is not in the document]
Comment (lwtdev) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/test/java/org/apache/shardingsphere/sql/parser/sql/common/extractor/TableExtractorTest.java:65]: Here I follow AAA(Arrange, Act, Assert) pattern, so there are two blank lines.

[the remark this answers is not in the document]
Comment (lwtdev) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/test/java/org/apache/shardingsphere/sql/parser/sql/common/extractor/TableExtractorTest.java:65]: But I found that it doesn’t seem to be the case in the project, so I will keep with the project and delete these blank lines

**Comment by lwtdev** [comment]
> @lwtdev Now, there is no UT for expr, are you interested in fix it? The relevant issue is #7880

@jingshanglu I don't known about what should to do about the work  `UT for expr`, Which ‘expr’ need to add test cases?

File: shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-mysql/src/main/antlr4/imports/mysql/DMLStatement.g4 [hunk 3]
```diff
@@ -108,10 +108,6 @@ multipleTablesClause
     : tableAliasRefList FROM tableReferences | FROM tableAliasRefList USING tableReferences
     ;
 
-multipleTableNames
```
[the remark this answers is not in the document]
Comment (lwtdev) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-mysql/src/main/antlr4/imports/mysql/DMLStatement.g4:111]: Yes, it's useless. More detailed discussion here #7918

File: shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/mysql/visitor/statement/impl/MySQLStatementSQLVisitor.java [hunk 4]
```diff
@@ -544,7 +545,17 @@ public ASTNode visitQueryExpressionParens(final QueryExpressionParensContext ctx
 
     @Override
     public ASTNode visitLockClauseList(final LockClauseListContext ctx) {
-        return new LockSegment(ctx.getStart().getStartIndex(), ctx.getStop().getStopIndex());
+        LockSegment lockSegment = new LockSegment(ctx.getStart().getStartIndex(), ctx.getStop().getStopIndex());
```
[the remark this answers is not in the document]
Comment (lwtdev) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-dialect/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/mysql/visitor/statement/impl/MySQLStatementSQLVisitor.java:548]: OK, I will modify this.

File: shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/dml/predicate/LockSegment.java [hunk 5]
```diff
@@ -32,4 +36,7 @@
     private final int startIndex;
 
     private final int stopIndex;
+
+    @Setter
+    private List<SimpleTableSegment> forTables;
```
[the remark this answers is not in the document]
Comment (lwtdev) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/dml/predicate/LockSegment.java:41]: In fact, I thought pre-initialized it at the beginning. But the `forTables` is empty  most time. And I also found some code example in Shardingshpere project that keep list as null default.  So I am also a little confused at it.
But I also think pre-initialization is easier to read and use. 

I want to keep the same style with the project, so what is the style of the project? 😀

[the remark this answers is not in the document]
Comment (lwtdev) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/dml/predicate/LockSegment.java:41]: I also tried this style before 😀
```java
private final Optional<List< SimpleTableSegment >> tables = Optional.empty();
```

[the remark this answers is not in the document]
Comment (strongduanmu) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/dml/predicate/LockSegment.java:41]: > In fact, I thought pre-initialized it at the beginning. But the `forTables` is empty most time. And I also found some code example in Shardingshpere project that keep list as null default. So I am also a little confused at it.
> But I also think pre-initialization is easier to read and use.
> 
> I want to keep the same style with the project, so what is the style of the project? 😀

@lwtdev I found two common coding styles, one is to use construction method to initialize the collection, such as the `values` ​​variable in `InsertValuesSegment`, and the other is to pre-initialize directly in the class, such as the `variableAssigns` variable in `SetStatement`. 😀

[the remark this answers is not in the document]
Comment (lwtdev) [reply shardingsphere-sql-parser/shardingsphere-sql-parser-statement/src/main/java/org/apache/shardingsphere/sql/parser/sql/common/segment/dml/predicate/LockSegment.java:41]: @strongduanmu I will pre-initialize the `tables` 😀~

**Review by strongduanmu — APPROVED** [review]

**Review by jingshanglu — APPROVED** [review]

**Comment by jingshanglu** [comment]
> > @lwtdev Now, there is no UT for expr, are you interested in fix it? The relevant issue is #7880
> 
> @jingshanglu I don't known what should to do about the work `UT for expr`, Which ‘expr’ need to add test cases?

@lwtdev There is `WhereSegment`  in selectStatement like this:
```
public abstract class SelectStatement extends AbstractSQLStatement implements DMLStatement {
    
    private ProjectionsSegment projections;
    
    private TableSegment from;
    
    private WhereSegment where;
    
    private GroupBySegment groupBy;
    
    private OrderBySegment orderBy;
}
// WhereSegment contain expr
public final class WhereSegment implements SQLSegment {
    
    private final int startIndex;
    
    private final int stopIndex;
    
    private final ExpressionSegment expr;
}
```
But in our test framework, there is no assertion of expr, like this:
```
public final class WhereClauseAssert {
    
    /**
     * Assert actual where segment is correct with expected where clause.
     * 
     * @param assertContext assert context
     * @param actual actual where segment
     * @param expected expected where clause
     */
    public static void assertIs(final SQLCaseAssertContext assertContext, final WhereSegment actual, final ExpectedWhereClause expected) {
//        TODO support expr in where assert
    }
}
```
just like `assertLockClause()` that you added.

**Pull request merged**